core/drivers/pl022_spi.c

```diff
@@ -0,0 +1,532 @@
+#define SSPCR1_LBM_NO		SHIFT_U32(0, 0)
+
+#define SSPDR_DATA	SHIFT_U32(0xFFFF, 0)
+
+#define SSPSR_BSY	SHIFT_U32(1, 4)
+#define SSPSR_RNF	SHIFT_U32(1, 3)
+#define SSPSR_RNE	SHIFT_U32(1, 2)
+#define SSPSR_TNF	SHIFT_U32(1, 1)
+#define SSPSR_TFE	SHIFT_U32(1, 0)
+
+#define SSPCPSR_CPSDVR	SHIFT_U32(0xFF, 0)
+
+#define SSPIMSC_TXIM	SHIFT_U32(1, 3)
+#define SSPIMSC_RXIM	SHIFT_U32(1, 2)
+#define SSPIMSC_RTIM	SHIFT_U32(1, 1)
+#define SSPIMSC_RORIM	SHIFT_U32(1, 0)
+
+#define SSPRIS_TXRIS	SHIFT_U32(1, 3)
+#define SSPRIS_RXRIS	SHIFT_U32(1, 2)
+#define SSPRIS_RTRIS	SHIFT_U32(1, 1)
+#define SSPRIS_RORRIS	SHIFT_U32(1, 0)
+
+#define SSPMIS_TXMIS	SHIFT_U32(1, 3)
+#define SSPMIS_RXMIS	SHIFT_U32(1, 2)
+#define SSPMIS_RTMIS	SHIFT_U32(1, 1)
+#define SSPMIS_RORMIS	SHIFT_U32(1, 0)
+
+#define SSPICR_RTIC		SHIFT_U32(1, 1)
+#define SSPICR_RORIC	SHIFT_U32(1, 0)
+
+#define SSPDMACR_TXDMAE	SHIFT_U32(1, 1)
+#define SSPDMACR_RXDMAE	SHIFT_U32(1, 0)
+
+#define SSPPeriphID0_PartNumber0	SHIFT_U32(0xFF, 0) /* 0x22 */
+#define SSPPeriphID1_Designer0		SHIFT_U32(0xF, 4) /* 0x1 */
+#define SSPPeriphID1_PartNumber1	SHIFT_U32(0xF, 0) /* 0x0 */
+#define SSPPeriphID2_Revision		SHIFT_U32(0xF, 4)
+#define SSPPeriphID2_Designer1		SHIFT_U32(0xF, 0) /* 0x4 */
+#define SSPPeriphID3_Configuration	SHIFT_U32(0xFF, 0) /* 0x00 */
+
+#define SSPPCellID_0	SHIFT_U32(0xFF, 0) /* 0x0D */
+#define SSPPCellID_1	SHIFT_U32(0xFF, 0) /* 0xF0 */
+#define SSPPPCellID_2	SHIFT_U32(0xFF, 0) /* 0x05 */
+#define SSPPPCellID_3	SHIFT_U32(0xFF, 0) /* 0xB1 */
+
+#define MASK_32 0xFFFFFFFF
+#define MASK_28 0xFFFFFFF
+#define MASK_24 0xFFFFFF
+#define MASK_20 0xFFFFF
+#define MASK_16 0xFFFF
+#define MASK_12 0xFFF
+#define MASK_8 0xFF
+#define MASK_4 0xF
+/* spi register masks */
+
+#define SSP_CPSDVR_MAX	254
+#define SSP_CPSDVR_MIN	2
+#define SSP_SCR_MAX		255
+#define SSP_SCR_MIN		0
+
+enum pl022_data_size {
+	PL022_DATA_SIZE4 = 0x3,
+	PL022_DATA_SIZE5,
+	PL022_DATA_SIZE6,
+	PL022_DATA_SIZE7,
+	PL022_DATA_SIZE8 = SSPCR0_DSS_8BIT,
+	PL022_DATA_SIZE9,
+	PL022_DATA_SIZE10,
+	PL022_DATA_SIZE11,
+	PL022_DATA_SIZE12,
+	PL022_DATA_SIZE13,
+	PL022_DATA_SIZE14,
+	PL022_DATA_SIZE15,
+	PL022_DATA_SIZE16 = SSPCR0_DSS_16BIT
+};
+
+enum pl022_spi_mode {
+	PL022_SPI_MODE0 = SSPCR0_SPO0 | SSPCR0_SPH0, /* 0x00 */
+	PL022_SPI_MODE1 = SSPCR0_SPO0 | SSPCR0_SPH1, /* 0x80 */
+	PL022_SPI_MODE2 = SSPCR0_SPO1 | SSPCR0_SPH0, /* 0x40 */
+	PL022_SPI_MODE3 = SSPCR0_SPO1 | SSPCR0_SPH1  /* 0xC0 */
+};
+
+static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat,
+	uint8_t *rdat, uint32_t num_txpkts, uint32_t *num_rxpkts)
+{
+	uint32_t i, j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	for (i = 0; i < num_txpkts; i++) {
+		if (read32(pd->base + SSPSR) & SSPSR_TNF) {
+			/* tx 1 packet */
+			write8(wdat[i], pd->base + SSPDR);
+
+			/* rx 1 packet */
+			if (read32(pd->base + SSPSR) & SSPSR_RNE)
+				rdat[j++] = read8(pd->base + SSPDR);
```

Comment:
The transfer by itself is always one packet/word in and one out, but since SPI doesn't specify rules for number of packets transmitted and/or received (does it?), the code was written with the assumption that there's a possibility to receive more packets than sent, and decisions regarding buffer size left to the caller assumed to have that knowledge, but yes, this does lead to a possibility of overflow.

What about
1. Read only as many packets as transmitted, or
2. Add a parameter to specify the length of the buffer, and only read up to its size, or
3. Other better suggestions?